libfwupdplugin/fu-sbatlevel-section-firmware.c

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright (C) 2023 Canonical Ltd.
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#define G_LOG_DOMAIN "FuFirmware"
+
+#include "config.h"
+
+#include "fu-bytes.h"
+#include "fu-csv-firmware.h"
+#include "fu-mem.h"
+#include "fu-sbatlevel-section-firmware.h"
+#include "fu-sbatlevel-section-struct.h"
+
+G_DEFINE_TYPE(FuSbatlevelSectionFirmware, fu_sbatlevel_section_firmware, FU_TYPE_FIRMWARE);
+
+static gboolean
+fu_sbatlevel_section_add_entry(FuFirmware *firmware,
+			       GBytes *fw,
+			       const guint8 *buf,
+			       gsize bufsz,
+			       gsize entry_offset,
+			       const gchar *entry_name,
+			       guint64 entry_idx,
+			       FwupdInstallFlags flags,
+			       GError **error)
+{
+	g_autoptr(FuFirmware) entry_fw = NULL;
+	g_autoptr(GBytes) entry_blob = NULL;
+	gsize entry_size;
+
+	/* look for the null terminator */
+	for (entry_size = 0; ((entry_offset + entry_size) < bufsz); ++entry_size) {
+		if (buf[entry_offset + entry_size] == 0)
+			break;
+	}
+
+	entry_fw = fu_csv_firmware_new();
+	fu_csv_firmware_add_column_id(FU_CSV_FIRMWARE(entry_fw), "$id");
+	fu_csv_firmware_add_column_id(FU_CSV_FIRMWARE(entry_fw), "component_generation");
+	fu_csv_firmware_add_column_id(FU_CSV_FIRMWARE(entry_fw), "date_stamp");
+
+	fu_firmware_set_idx(entry_fw, entry_idx);
+	fu_firmware_set_id(entry_fw, entry_name);
+	fu_firmware_set_offset(entry_fw, entry_offset);
+	entry_blob = fu_bytes_new_offset(fw, entry_offset, entry_size, error);
+	if (entry_blob == NULL)
+		return FALSE;
+	if (!fu_firmware_add_image_full(firmware, entry_fw, error))
+		return FALSE;
+	if (!fu_firmware_parse(entry_fw, entry_blob, flags, error))
+		return FALSE;
+
+	return TRUE;
+}
+
+static gboolean
+fu_sbatlevel_section_firmware_parse(FuFirmware *firmware,
+				    GBytes *fw,
+				    gsize offset,
+				    FwupdInstallFlags flags,
+				    GError **error)
+{
+	const guint8 *buf = NULL;
+	gsize bufsz;
+	gsize header_offset = offset + FU_STRUCT_SBAT_LEVEL_SECTION_HEADER_OFFSET_PREVIOUS;
+	guint32 previous_addr;
+	guint32 latest_addr;
+	g_autoptr(GByteArray) st = NULL;
+
+	buf = g_bytes_get_data(fw, &bufsz);
+
+	st = fu_struct_sbat_level_section_header_parse(buf, bufsz, offset, error);
+	if (st == NULL)
+		return FALSE;
+
+	previous_addr = fu_struct_sbat_level_section_header_get_previous(st);
+	latest_addr = fu_struct_sbat_level_section_header_get_latest(st);
```

Comment:
i think move this line to just above the 2nd fu_sbatlevel_section_add_entry block -- in fact you could just use fu_struct_sbat_level_section_header_get_latest() directly rather than a temp variable.

Reply:
I have moved it. The name looked long when inline.